File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    const termMap = terms.reduce((obj, term) => {
        const [str, num] = term.split(' ')
        obj[str] = Number(num) // ex) { 'A' : 6 }
        return obj
    }, {})
    
    return privacies.reduce((arr, privacy, index) => {
        const [date, str] = privacy.split(' ')
        const expireTime = getTime(date) + termMap[str] * 28
        if (expireTime <= getTime(today)) arr.push(index+1)
        return arr
    }, [])
}

function getTime(date) {
    const [year, month, day] = date.split('.').map(Number)
    return year * 12 * 28 + month * 28 + day
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        Map<String, Integer> termMap = new HashMap<>();
        for (String term : terms) {
            String[] t = term.split(" ");
            termMap.put(t[0], Integer.parseInt(t[1]));
        }
       
        List<Integer> answer = new ArrayList<>();
        int todayTime = getTime(today);
            
        for (int i=0; i<privacies.length; i++) {
            String[] dateAndTerm = privacies[i].split(" ");
            if (getTime(dateAndTerm[0]) + termMap.get(dateAndTerm[1]) * 28 <= todayTime) answer.add(i+1);
        }
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
    
    private int getTime(String date) {
        String[] dates = date.split("\\.");
        return Integer.parseInt(dates[0]) * 12 * 28
            + Integer.parseInt(dates[1]) * 28
            + Integer.parseInt(dates[2]);
    }
}